test/src/com/xilinx/rapidwright/util/TestBlackBoxPopulator.java

```diff
@@ -0,0 +1,93 @@
+
+import com.xilinx.rapidwright.design.Design;
+import com.xilinx.rapidwright.design.Module;
+import com.xilinx.rapidwright.design.Net;
+import com.xilinx.rapidwright.interchange.DeviceResources;
+import com.xilinx.rapidwright.support.RapidWrightDCP;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestBlackBoxPopulator {
+
+    private static String topDCPName = RapidWrightDCP.getString("hwct.dcp");
+    private static String cellDCPName = RapidWrightDCP.getString("hwct_pr1.dcp");
+    private static String cellAnchor = "INT_X0Y60";
+    private static List<Pair<String, String>> targets = new ArrayList<Pair<String, String>>()
+    {{
+        add(new Pair<>("hw_contract_pr0", "INT_X0Y120"));
+        add(new Pair<>("hw_contract_pr1", "INT_X0Y60"));
+        add(new Pair<>("hw_contract_pr2", "INT_X0Y0"));
+    }};
+
+    private int numPIPs(Design d) {
+        int PIPCount = 0;
+        for (Net n : d.getNets()) {
+            PIPCount += n.getPIPs().size();
+        }
+        return PIPCount;
+    }
+
+    @Test
+    void testRelocateModuleInsts() {
+        Design top = Design.readCheckpoint(topDCPName);
+        int numCellTop = top.getCells().size();
+        int numVccNetTop = (top.getVccNet() == null) ? 0 : 1;
+        int numGndNetTop = (top.getGndNet() == null) ? 0 : 1;
+        int numSignalNetTop = top.getNets().size() - numVccNetTop - numGndNetTop;
+        int numPIPTop = numPIPs(top);
+
+        Design template = Design.readCheckpoint(cellDCPName);
+        int numVccNetTemplate = (template.getVccNet() == null) ? 0 : 1;
+        int numGndNetTemplate = (template.getGndNet() == null) ? 0 : 1;
+        int numSignalNetTemplate = template.getNets().size() - numVccNetTemplate - numGndNetTemplate;
+        int numPIPTemplate = numPIPs(template);
+
+        Module mod = new Module(template, false);
+        BlackboxPopulator.relocateModuleInsts(top, mod, cellAnchor, targets);
+
+        Assertions.assertEquals(targets.size()*template.getCells().size()+numCellTop, top.getCells().size()
+                ,"Wrong number of cells!");
+
+        int numExpectedNets = targets.size()*numSignalNetTemplate + numSignalNetTop
+                              + Math.max(numVccNetTop, numVccNetTemplate) + Math.max(numGndNetTop, numGndNetTemplate);
+        Assertions.assertEquals(numExpectedNets, top.getNets().size(),"Wrong number of nets!");
+
+        Assertions.assertEquals(targets.size()*numPIPTemplate + numPIPTop, numPIPs(top),"Wrong number of PIPs!");
+    }
+    @Test
+    void testOutputDCP() {
+        Design top = Design.readCheckpoint(topDCPName);
+        Design template = Design.readCheckpoint(cellDCPName);
+        Module mod = new Module(template, false);
+        BlackboxPopulator.relocateModuleInsts(top, mod, cellAnchor, targets);
+
+        top.writeCheckpoint("output.dcp");
+        Design.readCheckpoint("output.dcp");
+    }
```

Comment:
Is this necessary?

Reply:
This is to catch the library error we saw yesterday.  If there is a check for that somewhere else, this could be remove.

Reply:
Writing and reading a DCP makes this a slow test. Can you check if the library of the problematic cell is "work" instead? 
If not, please leave a comment as to what this tests, and take a look at other tests to see how you can write the output to a temporary directory instead.

Reply:
change the test to check the library names and add comment for the test.